File: examples/erlang/user-client/test/user_http_mock.erl
-module(user_http_mock).
-export([request/4]).

request(get, {Url, _Headers, Mime, _Body}, [], [{body_format, binary}]) ->
    case {Url, Mime} of
        {"https://api.example/items", "application/json"} ->
            ok_response(200, [{"etag", "\"v1\""}], <<"{\"ok\":true}">>);
        {"https://api.example/fail", _} ->
            {error, timeout};
        {Url, "application/octet-stream"} ->
            case string:prefix(Url, "https://api.example/items?") of
                nomatch -> {error, {unexpected_request, Url, Mime}};
                _ -> ok_response(200, [], <<>>)
            end;
        _ ->
            {error, {unexpected_request, Url, Mime}}
    end;
request(_Method, _Req, _HttpOpts, _Opts) ->
    {error, unexpected_method}.

ok_response(Status, Headers, Body) ->
    {ok, {{http, Status, <<"OK">>}, Headers, Body}}.

File: examples/elixir/user-client/test/support/user_http_mock.ex
defmodule UserHttpMock do
  @moduledoc false

  def request(req_opts) do
    method = Keyword.fetch!(req_opts, :method)
    url = Keyword.fetch!(req_opts, :url)
    params = Keyword.get(req_opts, :params, %{})
    headers = Keyword.get(req_opts, :headers, [])

    case {method, url, params, headers} do
      {:get, "https://api.example/items", params, headers} when map_size(params) == 0 ->
        case List.keyfind(headers, "Content-Type", 0) do
          {"Content-Type", "application/json"} ->
            {:ok, %{status: 200, headers: [{"etag", "\"v1\""}], body: ~s({"ok":true})}}

          _ ->
            {:error, {:unexpected_request, req_opts}}
        end

      {:get, "https://api.example/items", %{"verbose" => "true"}, _} ->
        {:ok, %{status: 200, headers: [], body: ""}}

      {:get, "https://api.example/fail", _, _} ->
        {:error, :timeout}

      _ ->
        {:error, {:unexpected_request, req_opts}}
    end
  end
end
